### ISSUE TITLE
Not possible to add link with fragment identifier

### ISSUE THREAD

**Issue by sirex** [wagtail/wagtail#1049]
Currently it is not possible to add [fragment identifier](http://en.wikipedia.org/wiki/Fragment_identifier) to Wagtail rich text links:

https://github.com/torchbox/wagtail/blob/master/wagtail/wagtailcore/rich_text.py#L104

**Comment by davecranwell** [comment]
Please could you provide a little more information about what you're expecting here? Do you mean when adding Internal links? External ones? 

Have you also tried this in the latest version of wagtail (1.0rc2) where fragment identifiers certainly are supported in _external_ links?

**Comment by sirex** [comment]
@davecranwell the issue is with `a` tag with `linktype="page"`, this tag is interpreted by Wagtail for rich text fields:

```html
<a linktype="page" id="42">hello world</a>
```

How can I add `#some-ref` here?

**Comment by davecranwell** [comment]
Ah ok, internal links. Yes at present wagtail doesn't support creating internal links which include a fragment identifier, whether via the rich text editor or the PageChooserPanel widget.

**Comment by sirex** [comment]
I guess it is not an easy task to add internal links support to the rich text editor and the PageChooserPanel widget.

But maybe adding support just for `a` tag parser would be a good start? For example this:

```html
<a linktype="page" id="42" ref="some-ref">hello world</a>
```

Could be converted to:

```html
<a href="/some/page/#some-ref">hello world</a>
```

Of course editing such links with the rich text editor would remove fragment identifierds, but in may case, that would be acceptable.

In my use case, texts from old website are automatically imported to Wagtail, but currently I can't preserve internal page links because of this limitation. So I have bunch of links that just point to the same page and not to the parts of same page.

**Comment by gasman** [comment]
Adding a new attribute for this in the DB representation seems like a good idea (although I'd probably call it `fragment` rather than `ref`). It would just need a few additions to `PageLinkHandler` in wagtailcore/rich_text.py - and if we get it to output a `data-fragment` attribute in the `for_editor` case, we'll be able to preserve the identifier in the round trip from the database to the editor and back again.

The one thing that's missing is the ability to insert such a link within the rich text editor. I'm undecided about the best way to handle this (we could just add a new text field in the link chooser UI, to enter a fragment ID - but that still relies on the template developer to insert `<a id="...">` elements in the appropriate places, and requires the editor to understand some low-level HTML details) - but as you point out, it's not necessary to solve that problem yet.

**Comment by timheap** [comment]
You could add a new `Page.available_fragments()` method that returns a list of fragments the page has. Subclasses could override this to provide their own. The default implementation would either return an empty list `[]` or raise a `NotImplementedException`. On the front end, when a page is chosen for a link, a drop down of available fragments could then be presented, falling back to a text box if the `available_fragments()` did not return anything.

**Comment by ababic** [comment]
Resurrecting this again :)

I would say that having something built into page types to automatically gather a list of valid 'fragments' is nice, but it's probably overkill. Fragments within a page are likely to change over time (in the project that bought this to our attention, most 'in-page anchors' are defined within StreamFields, so that's very likely), and there also may be a mix of anchors that are hardcoded into templates, as well as being defined in content... which makes things tough to consolidate if you have different team members working on things over time (as we do).

As an editing experience, I think it would be fine to just give a text box to specify the fragment part. If that was incorrect or became incorrect over time, at least the link would still get the person to the right page, which is the important thing.

**Comment by ababic** [comment]
One thing the above wouldn't cater for very well is 'linking to a fragment on the same page as the one being edited'... Could we allow a fragment to be specified without a page? Would it be a good idea to represent that differently in the database representation, do you think?

**Comment by colingreen** [comment]
I'm facing the same problem - content editors would like to link to section headings in pages using fragment links.

But, the "Add an external link" form doesn't allow anything that starts with '#'.

Would it just be a case of making `wagtail.forms.URLOrAbsolutePathValidator` accept fragments? I'm more than happy to submit a pull request if so.

**Comment by strindhaug** [comment]
Making external link accept anchorl links is really simple:

```diff
diff --git a/wagtail/wagtailadmin/forms.py b/wagtail/wagtailadmin/forms.py
index c19f6686d..ad645b151 100644
--- a/wagtail/wagtailadmin/forms.py
+++ b/wagtail/wagtailadmin/forms.py
@@ -28,7 +28,7 @@ from wagtail.wagtailcore.models import (
 class URLOrAbsolutePathValidator(validators.URLValidator):
     @staticmethod
     def is_absolute_path(value):
-        return value.startswith('/')
+        return value.startswith('/') or value.startswith('#')
 
     def __call__(self, value):
         if URLOrAbsolutePathValidator.is_absolute_path(value):
```
Though it should then probably be renamed to `URLOrAbsolutePathOrAnchorValidator.is_absolute_path_or_anchor()` 
;)

**Comment by colingreen** [comment]
@strindhaug - that's what I was thinking about.

I also thought that it might be better to just make sure the value is a well formed URI, rather than having to be absolute. Don't want to open a can of worms though!

**Comment by gasman** [comment]
The problem I have with this approach is the need to enter the fragment as an "external link", when it absolutely isn't one... on a technical level it's a simple and convenient way of achieving the goal we want, but from a UX perspective it's flat out wrong. (In the case of #901 we were able to justify it with some semantic sleight-of-hand by saying "well, those links are external _to Wagtail_..." - that doesn't work in this case.)

I'm not sure what the real non-hackish solution is here. At the back of my mind, the need for editors to work with anchors / fragments directly is raising an alarm bell - ideally editors should be shielded from the details of HTML, and should be working in the domain of pure "content". I'm inclined to say that it really calls for a new link type - alongside internal (page) link, external link, email and document, you'd have "link to section", where the concept of a section is something specific to that site (and might have a dedicated chooser interface, populated dynamically from the sections on the page). I know this sounds massively over-engineered when all you want to do is insert an anchor link, but I'd like to at least find a solution that takes us a step closer to complete logical correctness rather than away from it :-)

**Comment by colingreen** [comment]
@gasman - How about going the other way from "logical correctness" towards just syntactic correctness. The HTML spec allows any valid URI in an href attribute. It's fine to provide chooser UIs for things that are managed by Wagtail. But for anything else we shouldn't impose constraints for the sake of it – for example: The same code prevents us from entering `tel:` or even `git:` urls, which are perfectly valid and well understood.

**Comment by colingreen** [comment]
Totally agreed about a "non-hackish" solution though – I can see how this might take a bit more work.

**Comment by gasman** [comment]
@colingreen I think it's a good thing for Wagtail to be conservative by default in the URIs it accepts, and only allow protocols that have been specifically whitelisted in some way - we don't want to allow arbitrary `javascript:` URLs, for example. (Informally, I think it also serves as a useful brake on throw-everything-into-a-rich-text-field syndrome: are you _really_ going to be putting `tel:` links in the middle of prose text, or is that just obscuring the real issue, namely that your StaffMemberPage model is lacking a `phone_number` field and people are having to add it in as a bullet point in rich text instead...?)

Having said that, we could certainly do a better job of making the list of allowed URI schemes/formats more configurable - at the moment, fiddling with HTML whitelisting rules is probably the only option.

**Comment by strindhaug** [comment]
What I want anchor links for is to link to a footnotes/bibliography  in the bottom of the page. Making a custom block for this is useless, since they need to be inline in the text and the text will be wrapped in both a `<div class="rich-text">` and it's content in `<p>` tags or other block level elements.

I've found a hacky workaround that doesn't require me to modify wagtail: use a "magic" absolute url prefix `/ref/<numberhere>` and replace the `/ref/`prefix with `#ref-` (so that it matches the corresponding reference in the reference list block)  using javascript (or some other preprocessing).  But it would be much more elegant if i could specify the anchor directly and avoid having to pre-process  the links and reserving url paths for magic usage to make it work.

**Comment by gasman** [comment]
@strindhaug I believe that migrating to draft.js (#2409) should handle this nicely. By decoupling the content model from HTML, we'll be able to define "footnote" as a link type in its own right, rather than editors having to construct them manually with anchor links.

**Comment by colingreen** [comment]
Digging a little deeper – `URLOrAbsolutePathValidator` first checks that the value starts with `/` and passes without checking if it's actually a valid URI. If the value doesn't start with `/` it's validated as a URI. `wagtail.wagtailadmin.forms.

URLOrAbsolutePathValidator` extends `django.core.validators. URLValidator` which by default only passes URLs with http(s) & ftp(s) schemes. 

With these things in mind, I honestly don't see a problem with allowing strings that start with '#' - there's no chance of users entering anything dodgy.

The only difficult bit for me is tweaking the naming and UI to match the logic.

**Comment by rifuso** [comment]
It's been some years now with this issue open. May I suggest that we go with strindhaug's solution at least as a stopgap? Because the inability to use in-page anchor tags makes Wagtail unsuitable for many sites.

**Comment by gasman** [comment]
OK, I'll accept a PR for allowing `#foo` as an external link, as per @colingreen and @strindhaug's suggestion. I should note that this only addresses the specific case of linking to an anchor on the _current_ page, rather than an arbitrary page on the site (which is what this ticket was originally asking for) - but for that specific case, the 'external link' interface is actually a better fit for this, since we're just capturing an `href` value to be used directly, and making people navigate the page tree to select the current page would be redundant. It's just the "external link" heading that makes it _seem_ wrong... So yep, let's go with this.

**Comment by chosak** [comment]
A bit late to this but I wanted to chime in with another vote for the more general approach proposed by @colingreen [above](https://github.com/wagtail/wagtail/issues/1049#issuecomment-307071032). We've run into the limitation with `tel:` links many times, where a phone number is included as part of text ("Call 202-555-5555 for more information about foo") and it'd be very convenient to use an external link as opposed to manually inserting HTML into the rich text block.

**Comment by colingreen** [comment]
Ok, I’ll have a look at this next week then. I think it’s a good idea to just deal with fragments here. I’d like to be able to use `tel:` and `mailto:` URIs, but @chosak,  I think it would be better to validate them properly using Django validators like we currently do with `http:` URIs. Something we can come back to in another issue maybe?

**Comment by gasman** [comment]
I'll reopen #3776 in a moment to continue the discussion on `tel:` URLs there.

**Comment by thibaudcolas** [comment]
Here is a prototype implementation of anchor links based on the new editor's capabilities: https://github.com/thibaudcolas/wagtail_draftail_experiments/tree/master/wagtail_draftail_anchors

From the user's perspective, here's what it looks like:

![anchors](https://user-images.githubusercontent.com/877585/37855826-fdfa52a2-2ef9-11e8-90a4-16fff4134845.gif)

This implementation is very similar to the [official docs’ "stock" entity](http://docs.wagtail.io/en/v2.0/advanced_topics/customisation/extending_draftail.html#creating-new-entities).

If you are considering this for your project, what you need to know is that:

- Conceptually, this is NOT the same as other links in the rich text editor – it’s a separate custom entity called "ANCHOR", whose implementation is specific to Draft.js / Draftail.
- As such, you will need to implement your own anchor picker (a dropdown, a free-form text field, an autocomplete – up to you).
- A limitation of custom entities is that you cannot copy-paste them between multiple Draft.js fields.
- **If we were to implement this on Wagtail, it would probably be managed as a link, with the other links. _As such, if you decide to implement it the "custom" way, you're on your own._ It will work, but if a future Wagtail version gets support for anchors yours will still be separate, unless you were to migrate them to the new built-ins**.

About the implementation,

- All of the code in that project except `wagtail_draftail_anchors` is from the `wagtail start` template.
- The anchors are stored in rich text as `<a data-anchor="true" href="#my-anchor">`. I'm not sure whether the `data-anchor="true"` part is necessary.
- I haven't implemented editing (or removal) of existing anchors – just adding of new ones. Technically this is possible, but I don't think Wagtail exposes its `Tooltip` component for reuse yet so this would be quite a bit of Draft.js / React UI work.
- All of the client-side in this project relies on ES2015+ features, and won't work in IE11 for example.

**Comment by KatieMoll** [comment]
Just wanted to check-in and see how this enhancement is coming along.  We are looking forward to implementing it.  This looks like a very powerful feature to help users understand and quickly navigate to applicable content.

**Comment by quocduan** [comment]
@thibaudcolas i did fork from your repo and make some code for adding `anchor identifier` to richtext editor. Now we can create two things to make `in-page navigation` work.

1. Create `anchor identifier`: which is composed with `<a id="some-ref"`. This is the target portion of page we will be navigated to when click to some anchor marked by `href="#some-ref"`, for ex.
2. Create `anchor`

This is my repo: [](https://github.com/quocduan/wagtail_draftail_anchors.git)
PS: i've still not implement the features of **remove/edit** anchor/identifier yet.
Hope someone can do that!

**Comment by thibaudcolas** [comment]
@quocduan nice! This is exactly what I had in mind. Then you can set up targets to be placed either on text or as empty tags (`<span id="target"></span>`). And you can also add some validation that anchors only point to known targets (easy), and that targets really use unique identifiers (hard).

For the editing and removal, this is something a few people have asked about so I'll try to make a tutorial via my [Patreon](https://www.patreon.com/thibaud_colas). There are new built-in React components available in Wagtail 2.1 that should make this much easier (`Tooltip`, `Portal`, and a few others).

---

@KatieMoll as far as I know there is no one working on implementing this directly within Wagtail at the moment. We've made it easier to implement such features (e.g. "Footnote" links) with the new editor, and some people probably are using those in their projects.

If anyone wants to tackle this I'm happy to help. First step is probably to figure out whether it's enough to change the existing link choosers' validation and UI (https://github.com/wagtail/wagtail/issues/1049#issuecomment-307404017) to support fragments, or whether there should be a dedicated link type with its own UI for this. Either way, exploring this as a reusable third-party package like @quocduan did with `wagtail_draftail_anchors` will help a lot.

**Comment by thibaudcolas** [comment]
Hey @quocduan, let's keep this issue about fragment links please. You can join the [#draftail channel on Wagtail’s Slack](https://github.com/wagtail/wagtail/wiki/Slack) if you want to discuss other potential extensions to the editor. For code blocks, have a look at https://github.com/FlipperPA/wagtailcodeblock/issues/13.

PS: thanks for sharing your blog!

**Comment by victoriachan** [comment]
> OK, I'll accept a PR for allowing `#foo` as an external link, as per @colingreen and @strindhaug's suggestion. I should note that this only addresses the specific case of linking to an anchor on the _current_ page, rather than an arbitrary page on the site (which is what this ticket was originally asking for) - but for that specific case, the 'external link' interface is actually a better fit for this, since we're just capturing an `href` value to be used directly, and making people navigate the page tree to select the current page would be redundant. It's just the "external link" heading that makes it _seem_ wrong... So yep, let's go with this.

Agree that "external link" heading seems wrong. What about in addition to "Internal link | External link | Email link" options, we just add an "Anchor link" option which actually is almost identical to External link except it accepts and only accepts anchor links (for current page)?

**Comment by dominhtri1995** [comment]
+1 really need this

**Comment by thibaudcolas** [comment]
For anyone looking into this, it might also be interesting to look at the following StackOverflow question: https://stackoverflow.com/questions/49415788/fragment-id-linking-in-wagtails-rich-text-content.

Although this issue initially only was about adding links, there are two related but separate problems to solve here:

- Adding links that point to anchors / fragment identifiers in rich text
- Adding anchors to content in rich text

I don't think it's necessary to solve these two at the same time, but the "adding anchors" problem might inform the solution to "adding anchor links".

For adding anchors, I can think of a few potential solutions, that each have their set of advantages and drawbacks:

- Rewrite the rich text HTML when it gets rendered, injecting anchors in the relevant tags, based on the textual content of the tags (e.g. slug of text within heading levels).
- Customise the rich text HTML storage to inject anchors in the relevant tags, based on the textual content of the tags (e.g. slug of text within heading levels).
- Use custom StreamField blocks for content that needs anchors (e.g. headings), and generate the anchor fragment identifier from the block’s UUID (stable), or from the block text. Or have a field for users to enter their own fragment identifier.
- Use custom Draftail blocks for content that needs anchors (e.g. headings), and generate the anchor fragment identifier from the block's key (not so stable), or from the block text. Or have a field for users to enter their own fragment identifier.
- Use a custom Draftail entity to inject anchors anywhere in rich text, with the user providing the fragment identifier.

I think the main deciding factors here are:

- Whether anchors are better generated "automagically", or managed by end users
- How big of a problem is it if an anchor link breaks because the content changed

---

Again, we might not need to solve the whole problem there and now, and solutions like https://github.com/wagtail/wagtail/issues/1049#issuecomment-307404017 could be a good quick fix for this issue, but I thought it'd be useful to share more details about the general use case for those anchor links.

**Comment by johnraz** [comment]
@gasman : does https://github.com/wagtail/wagtail/issues/1049#issuecomment-339635487 still applies? Is it ok if I work on a PR to just pass validation when using `#some-fragment` as an external link?

That would solve the "Adding links that point to anchors / fragment identifiers in rich text" which is really what this issue is about.
I don't believe a dedicated "anchor/fragment link" Draftail feature would make much sense as it would just duplicate what can be done with existing "links".
Relabelling "external link" to "external / fragment link" in the UI would probably be enough, IMO.

On the other hand there is the question of "Adding anchors to content in rich text", and maybe it would help and make things clearer to open a new issue to discuss just that.
I played with https://github.com/thibaudcolas/wagtail_draftail_experiments/tree/master/wagtail_draftail_anchors last week, it basically does what's needed but is still missing a proper UI.
That new issue could be about improving the work started by @thibaudcolas .
Also I'd need guidance on what part of the existing client code could be re-used and extended (eg: there is a Tooltip component into the core, could that be re-used?) and I believe this would be better discussed in a separate issue.

I'd like to help push this forward but in order to do so, I'd like to do it step by step:

- Submit a PR to allow fragment in external links, and maybe relabel the external link, and get this very issue closed.
- Open a new issue about a feature to add fragment to richtext elements.
- Work on a PR to improve the work done by @thibaudcolas to solve and close that new issue.

Would that be a good plan?

**Comment by lb-** [comment]
Fixed in #5505  via https://github.com/wagtail/wagtail/commit/794d40b86b3d9c652c37acbc00163890077a78b5

**Comment by johnraz** [comment]
So cool to see the first part of the issue being fixed 🎉 
To followup on my previous comment I opened #5515.
